ed25519/src/sign.cpp

```diff
@@ -2,32 +2,31 @@
 #include "libtorrent/aux_/disable_warnings_push.hpp"
 
 #include "libtorrent/ed25519.hpp"
-#include "sha512.h"
+#include "libtorrent/hasher512.hpp"
 #include "ge.h"
 #include "sc.h"
 
+using namespace libtorrent;
 
 void ed25519_sign(unsigned char *signature, const unsigned char *message, size_t message_len, const unsigned char *public_key, const unsigned char *private_key) {
-    sha512_context hash;
+    hasher512 hash;
     unsigned char hram[64];
     unsigned char r[64];
     ge_p3 R;
 
-
-    sha512_init(&hash);
-    sha512_update(&hash, private_key + 32, 32);
-    sha512_update(&hash, message, message_len);
-    sha512_final(&hash, r);
+    hash.update({reinterpret_cast<char const*>(private_key) + 32, 32});
+    hash.update({reinterpret_cast<char const*>(message), message_len});
+    std::memcpy(r, hash.final().data(), 64);
 
     sc_reduce(r);
     ge_scalarmult_base(&R, r);
     ge_p3_tobytes(signature, &R);
 
-    sha512_init(&hash);
-    sha512_update(&hash, signature, 32);
-    sha512_update(&hash, public_key, 32);
-    sha512_update(&hash, message, message_len);
-    sha512_final(&hash, hram);
+    hash.reset();
+    hash.update({reinterpret_cast<char const*>(signature), 32});
+    hash.update({reinterpret_cast<char const*>(public_key), 32});
+    hash.update({reinterpret_cast<char const*>(message), message_len});
+    std::memcpy(hram, hash.final().data(), 64);
```

Comment:
a similar API was rejected by @arvidn in the first commit, it's this way here because this code was working with a more primitive API, but it should be modernized at some point in the future, it would be nice to avoid the double memcpy btw